### PR TITLE
V2 to v3 merge

### PR DESCRIPTION
**Pull request by ReneWerner87**
Bring changes from v2 branch to main (v3) branch.

## Summary by CodeRabbit

- **Documentation**
	- Updated contribution guidelines and pull request templates for clarity and consistency.
	- Revised documentation to reflect functionality changes, warnings, validations, and clarifications across various components.
- **New Features**
	- Introduced changes to handle redirects with query parameters more effectively.
	- Enhanced logging capabilities with panic triggers for specific log levels and inclusion of IP and error tags in logs.
- **Bug Fixes**
	- Fixed a syntax error in a script related to release events.
	- Corrected grammar in contribution guidelines.
- **Refactor**
	- Systematic update in context acquisition methods across the application and tests, shifting from `NewCtx` to `AcquireCtx`.
	- Simplified timeout handling logic by modifying function parameters and removing deprecated features.
- **Tests**
	- Updated test cases to reflect changes in context management, logging, and middleware configurations.
	- Removed outdated comments and adjusted request handling in test setups.
- **Chores**
	- Excluded specific linters from running on test files for improved code quality checks.

### CODE REVIEW

**Comment by ReneWerner87** [comment]
we need to take a close look at the pr
I hope I haven't forgotten anything
maybe someone can look over the logger again

because we made pull requests for v2 and v3, it was difficult to merge, we would have had to transfer the v2 code to v3 first and then make the v3 adjustment

now git had overwritten the v3 customization with the changes from v2, because they were new for him -> hopefully I resolved it correctly

File: .github/workflows/benchmark.yml
Comment (nickajacks1) [inline .github/workflows/benchmark.yml:5]: was removing main intentional?
same question for the other workflows

File: docs/api/middleware/logger.md
Comment (nickajacks1) [inline docs/api/middleware/logger.md:119]: this line uses a tab but the others use 4 spaces

File: go.mod
Comment (nickajacks1) [inline go.mod:6]: You should be able to remove these with go mod tidy

File: middleware/healthcheck/healthcheck_test.go
Comment (nickajacks1) [inline middleware/healthcheck/healthcheck_test.go:21]: redundant assert

File: middleware/logger/logger_test.go
Comment (nickajacks1) [inline middleware/logger/logger_test.go:150]: I think this was intended to return an error (as opposed to type fakeOutput around line 630).

File: .github/workflows/benchmark.yml
Comment (ReneWerner87) [reply .github/workflows/benchmark.yml:5]: no thx

File: middleware/healthcheck/healthcheck_test.go
Comment (ReneWerner87) [reply middleware/healthcheck/healthcheck_test.go:21]: after the golangci hints i changed it via search and replace 
thx for the hint, i will remove it

File: middleware/logger/logger_test.go
Comment (ReneWerner87) [reply middleware/logger/logger_test.go:150]: sure

File: ctx_test.go
Comment (sixcolors) [inline ctx_test.go:1367]: To make this pass change:
```suggestion
		NameWithDefault  string `json:"name2" xml:"Name2" form:"name2" cookie:"name2" query:"name2" params:"name2" header:"name2"`
		ClassWithDefault int    `json:"class2" xml:"Class2" form:"class2" cookie:"class2" query:"class2" params:"class2" header:"class2"`
```

File: ctx_test.go
Comment (sixcolors) [inline ctx_test.go:1366]: to make this pass:

```suggestion
		NameWithDefault  string `json:"name2" xml:"Name2" form:"name2" cookie:"name2" query:"name2" params:"name2" header:"name2"`
		ClassWithDefault int    `json:"class2" xml:"Class2" form:"class2" cookie:"class2" query:"class2" params:"class2" header:"class2"`
```

**Comment by sixcolors** [comment]
I think https://github.com/gofiber/fiber/blob/9c0dc5ee5e97620e40543727274a34f6c3e47225/middleware/logger/logger_test.go#L166 should be `require.EqualValues(t, 1, *o)`

**Comment by sixcolors** [comment]
> I think
> 
> https://github.com/gofiber/fiber/blob/9c0dc5ee5e97620e40543727274a34f6c3e47225/middleware/logger/logger_test.go#L166
> 
> should be `require.EqualValues(t, 1, *o)`

same for https://github.com/gofiber/fiber/blob/9c0dc5ee5e97620e40543727274a34f6c3e47225/middleware/logger/logger_test.go#L181

**Comment by sixcolors** [comment]
I'm also seeing issues with Bind in the following tests:

```bash
--- FAIL: Test_Ctx_Parsers (0.02s)
    --- FAIL: Test_Ctx_Parsers/QueryParser (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1028379b4]

goroutine 631 [running]:
testing.tRunner.func1.2({0x102a580c0, 0x102efa8e0})
        /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1634 +0x33c
panic({0x102a580c0?, 0x102efa8e0?})
        /opt/homebrew/Cellar/go/1.22.0/libexec/src/runtime/panic.go:770 +0x124
github.com/gofiber/fiber/v3.(*DefaultCtx).Request(0x0?)
        /Users/sixcolors/Documents/GitHub/fiber/ctx.go:385 +0x4
github.com/gofiber/fiber/v3.Test_Ctx_Parsers.func7.1({0x102aff4d0, 0x140003d9508}, 0x14000863a70)
        /Users/sixcolors/Documents/GitHub/fiber/ctx_test.go:1426 +0x30
github.com/gofiber/fiber/v3.Test_Ctx_Parsers.func1(0x1400048da00, 0x102ae56a0)
        /Users/sixcolors/Documents/GitHub/fiber/ctx_test.go:1375 +0xac
github.com/gofiber/fiber/v3.Test_Ctx_Parsers.func7(0x1400048da00)
        /Users/sixcolors/Documents/GitHub/fiber/ctx_test.go:1425 +0x44
testing.tRunner(0x1400048da00, 0x1400033c6a0)
        /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 291
        /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1742 +0x318
FAIL    github.com/gofiber/fiber/v3     10.545s
```

File: ctx_test.go
Comment (sixcolors) [inline ctx_test.go:1449]:
```suggestion
func Test_Ctx_Parsers(t *testing.T) {
	t.Parallel()
	// setup

	type TestStruct struct {
		Name             string
		Class            int
		NameWithDefault  string `json:"name2" xml:"Name2" form:"name2" cookie:"name2" query:"name2" params:"name2" header:"name2"`
		ClassWithDefault int    `json:"class2" xml:"Class2" form:"class2" cookie:"class2" query:"class2" params:"class2" header:"class2"`
	}

	withValues := func(t *testing.T, app *App, actionFn func(c Ctx, testStruct *TestStruct) error) {
		t.Helper()
		c := app.AcquireCtx()
		defer app.ReleaseCtx(c)
		testStruct := new(TestStruct)

		require.NoError(t, actionFn(c, testStruct))
		require.Equal(t, "foo", testStruct.Name)
		require.Equal(t, 111, testStruct.Class)
		require.Equal(t, "bar", testStruct.NameWithDefault)
		require.Equal(t, 222, testStruct.ClassWithDefault)
	}

	t.Run("BodyParser:xml", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.SetContentType(MIMEApplicationXML)
			c.Request().SetBody([]byte(`<TestStruct><Name>foo</Name><Class>111</Class><Name2>bar</Name2><Class2>222</Class2></TestStruct>`))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("BodyParser:form", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.SetContentType(MIMEApplicationForm)
			c.Request().SetBody([]byte(`name=foo&class=111&name2=bar&class2=222`))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("BodyParser:json", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.SetContentType(MIMEApplicationJSON)
			c.Request().SetBody([]byte(`{"name":"foo","class":111,"name2":"bar","class2":222}`))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("BodyParser:multiform", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			body := []byte("--b\r\nContent-Disposition: form-data; name=\"name\"\r\n\r\nfoo\r\n--b\r\nContent-Disposition: form-data; name=\"class\"\r\n\r\n111\r\n--b\r\nContent-Disposition: form-data; name=\"name2\"\r\n\r\nbar\r\n--b\r\nContent-Disposition: form-data; name=\"class2\"\r\n\r\n222\r\n--b--")
			c.Request().SetBody(body)
			c.Request().Header.SetContentType(MIMEMultipartForm + `;boundary="b"`)
			c.Request().Header.SetContentLength(len(body))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("CookieParser", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.Set("Cookie", "name=foo;name2=bar;class=111;class2=222")
			return c.Bind().Cookie(testStruct)
		})
	})
	t.Run("QueryParser", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().URI().SetQueryString("name=foo&name2=bar&class=111&class2=222")
			return c.Bind().Query(testStruct)
		})
	})
	t.Run("ParamsParser", func(t *testing.T) {
		t.Skip("ParamsParser is not ready for v3")
		// t.Parallel()
		// app := New()
		// withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
		//	 c.route = &Route{Params: []string{"name", "name2", "class", "class2"}}
		//	 c.values = [30]string{"foo", "bar", "111", "222"}
		//	 return c.ParamsParser(testStruct)
		// })
	})
	t.Run("ReqHeaderParser", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.Add("name", "foo")
			c.Request().Header.Add("name2", "bar")
			c.Request().Header.Add("class", "111")
			c.Request().Header.Add("class2", "222")
			return c.Bind().Header(testStruct)
		})
	})
}
```

File: ctx_test.go
Comment (sixcolors) [inline ctx_test.go:1360]:
```suggestion
```

File: ctx_test.go
Comment (sixcolors) [inline ctx_test.go:1441]:
```suggestion
	t.Run("BodyParser:xml", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.SetContentType(MIMEApplicationXML)
			c.Request().SetBody([]byte(`<TestStruct><Name>foo</Name><Class>111</Class><Name2>bar</Name2><Class2>222</Class2></TestStruct>`))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("BodyParser:form", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.SetContentType(MIMEApplicationForm)
			c.Request().SetBody([]byte(`name=foo&class=111&name2=bar&class2=222`))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("BodyParser:json", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.SetContentType(MIMEApplicationJSON)
			c.Request().SetBody([]byte(`{"name":"foo","class":111,"name2":"bar","class2":222}`))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("BodyParser:multiform", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			body := []byte("--b\r\nContent-Disposition: form-data; name=\"name\"\r\n\r\nfoo\r\n--b\r\nContent-Disposition: form-data; name=\"class\"\r\n\r\n111\r\n--b\r\nContent-Disposition: form-data; name=\"name2\"\r\n\r\nbar\r\n--b\r\nContent-Disposition: form-data; name=\"class2\"\r\n\r\n222\r\n--b--")
			c.Request().SetBody(body)
			c.Request().Header.SetContentType(MIMEMultipartForm + `;boundary="b"`)
			c.Request().Header.SetContentLength(len(body))
			return c.Bind().Body(testStruct)
		})
	})
	t.Run("CookieParser", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().Header.Set("Cookie", "name=foo;name2=bar;class=111;class2=222")
			return c.Bind().Cookie(testStruct)
		})
	})
	t.Run("QueryParser", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
			c.Request().URI().SetQueryString("name=foo&name2=bar&class=111&class2=222")
			return c.Bind().Query(testStruct)
		})
	})
	t.Run("ParamsParser", func(t *testing.T) {
		t.Skip("ParamsParser is not ready for v3")
		// t.Parallel()
		// app := New()
		// withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
		//	 c.route = &Route{Params: []string{"name", "name2", "class", "class2"}}
		//	 c.values = [30]string{"foo", "bar", "111", "222"}
		//	 return c.ParamsParser(testStruct)
		// })
	})
	t.Run("ReqHeaderParser", func(t *testing.T) {
		t.Parallel()
		app := New()
		withValues(t, app, func(c Ctx, testStruct *TestStruct) error {
```

**Comment by sixcolors** [comment]
@gofiber/maintainers,

I've noticed a common issue that appears throughout the tests. Running tests in parallel using the `t.Parallel()` function can cause problems if the tests are not isolated correctly. When tests share a state, such as a common instance of `app`, running them in parallel can lead to race conditions. This means that one test could modify the state while another is reading it, leading to unpredictable results.

To fix this issue, we should refactor the tests and make sure that each test has its isolated state.

If we want to eliminate repetitive code for setup and teardown for each parallel test, we could consider using the testify/suite package.

For example, in the `Test_Ctx_Parsers` noted above, I also ran into issues with the `Test_Client_UserAgent`.

Example, Test_Client_UserAgent with a fix:
```go
func Test_Client_UserAgent(t *testing.T) {
	t.Parallel()

	setupApp := func(t *testing.T) *fasthttputil.InmemoryListener {
		t.Helper()
		ln := fasthttputil.NewInmemoryListener()
		app := New()
		app.Get("/", func(c Ctx) error {
			return c.Send(c.Request().Header.UserAgent())
		})
		go func() {
			require.NoError(t, app.Listener(ln, ListenConfig{
				DisableStartupMessage: true,
			}))
		}()
		time.Sleep(100 * time.Millisecond) // wait for server to start
		return ln
	}

	t.Run("default", func(t *testing.T) {
		t.Parallel()
		ln := setupApp(t)
		for i := 0; i < 5; i++ {
			a := Get("http://example.com")

			a.HostClient.Dial = func(_ string) (net.Conn, error) { return ln.Dial() }

			code, body, errs := a.String()

			require.Equal(t, StatusOK, code)
			require.Equal(t, defaultUserAgent, body)
			require.Empty(t, errs)
		}
	})

	t.Run("custom", func(t *testing.T) {
		t.Parallel()
		ln := setupApp(t)
		for i := 0; i < 5; i++ {
			c := AcquireClient()
			c.UserAgent = "ua"

			a := c.Get("http://example.com")

			a.HostClient.Dial = func(_ string) (net.Conn, error) { return ln.Dial() }

			code, body, errs := a.String()

			require.Equal(t, StatusOK, code)
			require.Equal(t, "ua", body)
			require.Empty(t, errs)
			ReleaseClient(c)
		}
	})
}
```

Please let me know if you have any questions or concerns.

**Comment by ReneWerner87** [comment]
@gofiber/maintainers @gofiber/contributors can you take another quick look at it before I merge
there are many changes and more eyes may see more

will merge it tonight, if no other findings speak against it

File: middleware/cors/utils.go
Comment (renanbastos93) [inline middleware/cors/utils.go:84]: Here we can reduce processing from the CPU.
```suggestion
	return true, strings.ToLower(parsedOrigin.Scheme + "://" + parsedOrigin.Host)
```

benchmark:
```
// BenchmarkBothLower-8   	16690329	        62.00 ns/op	      32 B/op	       1 allocs/op
// BenchmarkOneLower-8   	22892752	        50.97 ns/op	       0 B/op	       0 allocs/op
```

code
```go
func bothLower(scheme, domain string) string {
	return strings.ToLower(scheme) + "://" + strings.ToLower(domain)
}

func oneLower(scheme, domain string) string {
	return strings.ToLower(scheme + "://" + domain)
}

func BenchmarkBothLower(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = bothLower("https", "renanbastos.com.br")
	}
}

func BenchmarkOneLower(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = oneLower("https", "renanbastos.com.br")
	}
}
```

**Review by renanbastos93 — COMMENTED** [review]
Great job done by @gofiber/contributors and @gofiber/maintainers! Congratulations to all involved. I have a few comments to add. See you later!

File: middleware/cors/cors.go
Comment (gaby) [reply middleware/cors/cors.go:125]: I think it's a lint violation to use var inside a panic

Comment (gaby) [reply middleware/cors/cors.go:125]: We could call log.Panicf

File: go.mod
Comment (gaby) [reply go.mod:6]: I think those need to be removed @ReneWerner87

**Review by gaby — CHANGES_REQUESTED** [review]
Besides this, I dont see anything that sticks out that hasnt been mentioned before

File: middleware/cors/cors.go
Comment (renanbastos93) [reply middleware/cors/cors.go:125]: it does make sense @gaby

Comment (efectn) [reply middleware/cors/cors.go:125]: > We could call log.Panicf

I suppose we should adjust default logger for Panicf. It doesn't panic

Comment (gaby) [reply middleware/cors/cors.go:125]: Interesting... so log.Fatal() will call exit(). Panic is suppose to call panic

File: .github/scripts/sync_docs.sh
Comment (ReneWerner87) [reply .github/scripts/sync_docs.sh:29]: yes, thanks for the hint
it was OFF for v2 but should be active in v3

File: go.mod
Comment (ReneWerner87) [reply go.mod:6]: sure thx

File: middleware/cors/cors.go
Comment (ReneWerner87) [reply middleware/cors/cors.go:125]: i was wrong, it doesn´t -> but i will adjust it

File: .github/scripts/sync_docs.sh
Comment (nickajacks1) [inline .github/scripts/sync_docs.sh:30]:
```suggestion
if [ "$EVENT" == "release" ]; then
```

Comment (ReneWerner87) [reply .github/scripts/sync_docs.sh:30]: my fold

Comment (nickajacks1) [reply .github/scripts/sync_docs.sh:30]: The rabbit beat me to it!!!

**Review by renanbastos93 — COMMENTED** [review]
Awesome

File: ctx_test.go
Comment (ReneWerner87) [reply ctx_test.go:4816]: not really related to my merge changes -> we cover it in the next PR´s

File: ctx_test.go
Comment (gaby) [reply ctx_test.go:4951]: Those are separate app instances

File: middleware/proxy/proxy_test.go
Comment (gaby) [inline middleware/proxy/proxy_test.go:302]: This a big reduction in sleep time?

File: middleware/proxy/proxy_test.go
Comment (gaby) [inline middleware/proxy/proxy_test.go:309]: Same here

File: middleware/proxy/proxy_test.go
Comment (gaby) [inline middleware/proxy/proxy_test.go:312]: Same here

File: log/default_test.go
Comment (gaby) [inline log/default_test.go:94]: This can be done using:

```go
require.Panics(t, func() {
    Panicf("%s panic", work)
})
```

File: log/default_test.go
Comment (gaby) [inline log/default_test.go:60]: Same as below

Comment (gaby) [reply log/default_test.go:60]: https://pkg.go.dev/github.com/stretchr/testify/require#Panics

File: log/default_test.go
Comment (gaby) [reply log/default_test.go:94]: https://pkg.go.dev/github.com/stretchr/testify/require#Panics

File: middleware/proxy/proxy_test.go
Comment (ReneWerner87) [reply middleware/proxy/proxy_test.go:309]: the reduction for the sleep was intended 
and should make no difference

File: middleware/proxy/proxy_test.go
Comment (ReneWerner87) [reply middleware/proxy/proxy_test.go:312]: the reduction for the sleep was intended 
and should make no difference

File: middleware/proxy/proxy_test.go
Comment (ReneWerner87) [reply middleware/proxy/proxy_test.go:302]: the reduction for the sleep was intended 
and should make no difference

**Review by gaby — APPROVED** [review]
Lgtm

**Pull request merged**
